Add a number of improvements aimed at making the plugin more "automatic"

mikehoyle opened the pull request:
See commits for details. Main changes:

 - Add option to only display when in the Hunllef room itself.
 - Add option to automatically start the timer when first hit (rather than needing to click yourself)
 - Add automatic trample detection.

mikehoyle commented:
I'd also love to talk to you about your openness to a bigger change that updates the callout system as a whole.

So far I've noticed one of the bigger shortcomings of the plugin is that, with its disconnected "timer" approach, it can get desynced. If you miss one trample, or get trampled on the last hit of a cycle and it registers too late, things can get dicey.

My proposal would be to take a more active approach to the callouts. By reading Hunllef's animation ID's (like I do for trample detection), the exact timing of attacks and switches can be known throughout the fight. This would keep things working perfectly on time without any user input. Let me know if that's a change you're open to --  I think it would be a great improvement to this already very useful plugin!

mikehoyle commented:
Upon further reading of my last suggestion, attack detection via the animations would be against the rules. However I think the changes I've added in this PR should be acceptable, given they only contribute to or affect the timer.

Loze-Put commented:
Hi mikehoyle,

Thanks for your suggestions and contribution. I think your changes would be a good addition to the plugin, but I also think they are partially violating the [Jagex third party client guidelines](https://secure.runescape.com/m=news/third-party-client-guidelines?oldschool=1).

The guidelines clearly state that 

> "Any feature which adds additional visual or audio indicators of a boss mechanic except in cases where this is a manually triggered external helper."

is unacceptable. This plugin provides both a visual and audio indicator of boss mechanics. At this point this process is entirely manual (with the exception of real time passing). Your changes to start the timer automatically or add time to the timer upon a trample would be considered automation of this process. Thus I think both these changes are violating the guidelines.

In the past I have considered changing the plugin to use a timer based on game ticks to prevent it from desyncing. I think this would either be exactly at the edge of what is allowed, or perhaps already crossing it. Personally I've always used the original idea for this plugin as reference. Can I create a website (or another application) which mimics the plugins' functionality? If yes, it is alowed. If no, it is not. The current plugin could be replaced by having a website with the same functionality openend next to the game client. Changing the plugin to use in-game ticks, animations, hitsplats or anything like that could not be replicated by an external website. Thus I've never added any feature which would use this information.

I think your option to only show the plugin in the Hunllef room is acceptable, because it does not provide any information on the boss' attacks. It only shows and hides the plugin panel which requires full manual input. I think rather than implementing it as a second checkbox in the options it would be better to turn the current checkbox into a dropdown list of 3 options instead.

I'd like to hear your opinion on the matter. If you think I'm mistaken for these takes I'd also love to hear different opinions. I think in order to actually add these mechanics I'd like an approval from either a member of the RuneLite team or a staff member of Jagex.

-Loze Put

mikehoyle commented:
Going over the rules, unfortunately I think you're right. I didn't realize how much they restricted this type of plugin.

I've updated the PR to remove the commits that provide further automation, and cleaned up the new visibility option into a dropdown.

Loze-Put commented on src/main/java/com/hunllefhelper/HunllefHelperConfig.java:
```diff
@@ -9,13 +11,13 @@ public interface HunllefHelperConfig extends Config
-		keyName = "autoHide",
-		name = "Automatically hide",
-		description = "If checked, only show the plugin panel inside The Gauntlet"
+		keyName = "visibilityMode",
+		name = "Visibility Mode",
+		description = "Determines when the plugin panel is shown."
```
I think it would look nicer if this configuration is a bit more intuitive. 
```suggestion
		keyName = "panelVisibility",
		name = "Show plugin panel",
		description = "Determines when the plugin panel is shown."
```

Loze-Put commented on src/main/java/com/hunllefhelper/config/VisibilityMode.java:
```diff
@@ -0,0 +1,21 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum VisibilityMode
+{
+    AlwaysVisible("Always Visible"),
+    InsideGauntlet("Only Inside Gauntlet"),
+    OnlyInHunllefRoom("Only In Hunllef Room");
```
```suggestion
public enum PanelVisibility
{
    Always("Always"),
    InsideGauntlet("Inside Gauntlet"),
    AtHunllef("At Hunllef");
```

Loze-Put commented on src/main/java/com/hunllefhelper/HunllefHelperConfig.java:
```diff
@@ -1,5 +1,7 @@
 package com.hunllefhelper;
```
It's nice to have the config files in a seperate folder. Could you move this file to the config package as well?

Loze-Put commented on src/main/java/com/hunllefhelper/HunllefHelperPlugin.java:
```diff
@@ -220,6 +214,27 @@ private boolean isInTheGauntlet()
+			return false;
+		}
+
+		WorldPoint playerLocation = WorldPoint.fromLocalInstance(client, player.getLocalLocation());
+		int regionId = playerLocation.getRegionID();
+		if (regionId == REGION_ID_GAUNTLET_NORMAL || regionId == REGION_ID_GAUNTLET_CORRUPTED) {
+			int playerX = playerLocation.getRegionX();
+			int playerY = playerLocation.getRegionY();
+            return playerX >= HUNLLEF_ROOM_X_MIN && playerX <= HUNLLEF_ROOM_X_MAX
+                    && playerY >= HUNLLEF_ROOM_Y_MIN && playerY <= HUNLLEF_ROOM_Y_MAX;
+		}
+
+		return false;
```
```suggestion
		WorldPoint playerLocation = WorldPoint.fromLocalInstance(client, player.getLocalLocation());
		int regionId = playerLocation.getRegionID();
		
		if (regionId != REGION_ID_GAUNTLET_NORMAL && regionId != REGION_ID_GAUNTLET_CORRUPTED)
		{
			return;
		{
		
		int playerX = playerLocation.getRegionX();
		int playerY = playerLocation.getRegionY();
		
		return playerX >= HUNLLEF_ROOM_X_MIN && playerX <= HUNLLEF_ROOM_X_MAX
			&& playerY >= HUNLLEF_ROOM_Y_MIN && playerY <= HUNLLEF_ROOM_Y_MAX;
```

Loze-Put commented on src/main/java/com/hunllefhelper/HunllefHelperPlugin.java:
```diff
@@ -99,12 +93,13 @@ protected void shutDown() throws Exception
-		boolean isInInstance = isInTheGauntlet();
+		boolean isInInstance = config.visibilityMode() == VisibilityMode.OnlyInHunllefRoom
+				? isInHunllefRoom() : isInTheGauntlet();
 		if (isInInstance != wasInInstance)
```
I think these variables should be renamed. The `InInstace` part is logical when auto hiding based on being in the gauntlet instances. Perhaps `wasPanelVisible` and `shouldPanelBeVisible` or something similair are better names.

I'm also fine with leaving the names as they are. I think I will refactor the code a bit later to move part of the hiding logic to a separate component.

Loze-Put reviewed (commented):
Looks pretty good to me 👍

mikehoyle commented:
All comments are addressed. Thanks!

Loze-Put commented:
Thanks for your contribution! I will add a migration to automatically transfer the old configuration to the new one. After that I'll submit the changes to the RuneLite plugin hub so everyone can enjoy them ingame.
